src/SFA.DAS.ApprenticeFeedback.Application/Queries/GetApprenticeFeedbackResults/GetApprenticeFeedbackResultsQueryHandler.cs

```diff
@@ -0,0 +1,83 @@
+using SFA.DAS.ApprenticeFeedback.Domain.Configuration;
+using SFA.DAS.ApprenticeFeedback.Domain.Entities;
+using SFA.DAS.ApprenticeFeedback.Domain.Interfaces;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using static SFA.DAS.ApprenticeFeedback.Application.Queries.GetApprenticeFeedbackResult.GetApprenticeFeedbackResultsResult;
+
+namespace SFA.DAS.ApprenticeFeedback.Application.Queries.GetApprenticeFeedbackResult
+{
+    public class GetApprenticeFeedbackResultsQueryHandler : IRequestHandler<GetApprenticeFeedbackResultsQuery, GetApprenticeFeedbackResultsResult>
+    {
+        private readonly IApprenticeFeedbackResultContext _apprenticeFeedbackResultContext;
+        private readonly IProviderAttributeContext _providerAttributeContext;
+        private readonly IAttributeContext _attributeContext;
+        private readonly ApplicationSettings _settings;
+
+        public GetApprenticeFeedbackResultsQueryHandler(
+            IApprenticeFeedbackTargetContext apprenticeFeedbackTargetContext,
+            IApprenticeFeedbackResultContext apprenticeFeedbackResultContext,
+            IProviderAttributeContext providerAttributeContext,
+            IAttributeContext attributeContext,
+            IDateTimeHelper dateTimeHelper,
+            ApplicationSettings settings
+            )
+        {
+            _apprenticeFeedbackResultContext = apprenticeFeedbackResultContext;
+            _providerAttributeContext = providerAttributeContext;
+            _attributeContext = attributeContext;
+            _settings = settings;
+        }
+
+        public async Task<GetApprenticeFeedbackResultsResult> Handle(GetApprenticeFeedbackResultsQuery request, CancellationToken cancellationToken)
+        {
+            var result = new GetApprenticeFeedbackResultsResult();
+
+            var sprocResult = await _apprenticeFeedbackResultContext.GetFeedbackForProvidersAsync(request.Ukprns, _settings.ReportingMinNumberOfResponses, _settings.ReportingFeedbackCutoffMonths);
+
+            if (sprocResult.Any())
+            {
+                var ukprns = sprocResult.Select(u => u.Ukprn).Distinct().ToArray();
+                var attributes = _attributeContext.Entities.ToList();
+
+                foreach (var ukprn in ukprns)
+                {
+                    var ratings = sprocResult.Where(r => r.Ukprn == ukprn).GroupBy(grp => grp.ProviderRating).Select(grp => new { Rating = grp.Key, Count = grp.Count() });
+
+                    var feedbackResult = new UkprnFeedback();
+                    feedbackResult.Ukprn = ukprn;
+
+                    foreach (var rating in ratings)
+                    {
+                        if (!string.IsNullOrEmpty(rating.Rating))
+                        {
+                            feedbackResult.ProviderRating.Add(rating.Rating, rating.Count);
+                        }
+                    }
+
+                    // @ToDo: a more elegant way of doing this ?
```

Comment:
I think there should be a more elegant way to do this, rather than relying on the response from the query above, if we have the UKPRN, then we know they have feedback to be returned based on the UKPRN in the result, Is it worth crafting another stored procedure or view that does something like this?  to allow the database to do some heavy lifting?

`SELECT at.AttributeId, at.AttributeName, Count(pa.AttributeValue) as Total, CAST( CASE WHEN pa.AttributeValue = 1 THEN 'Agree' ELSE 'Disagree' END as nvarchar) as Selection
FROM ApprenticeFeedbackTarget aft 
JOIN ApprenticeFeedbackResult afr ON aft.Id = afr.ApprenticeFeedbackTargetId 
JOIN ProviderAttribute pa ON afr.Id = pa.ApprenticeFeedbackResultId
JOIN Attribute at ON at.AttributeId = pa.AttributeId
WHERE Ukprn = 10000446 and DateTimeCompleted > DATEADD(month,-12,GETUTCDATE())
GROUP By at.AttributeId, at.AttributeName, AttributeValue
Order by at.AttributeId, Selection`

Reply:
Moved to https://skillsfundingagency.atlassian.net/browse/QF-528